// primeng/esm5/codehighlighter/codehighlighter.js
import { NgModule, Directive, ElementRef, AfterViewInit } from '@angular/core';
import { CommonModule } from '@angular/common';
var CodeHighlighter = /** @class */ (function () {
    function CodeHighlighter(el) {
        this.el = el;
    }
    CodeHighlighter.prototype.ngAfterViewInit = function () {
        if (window['Prism']) {
            window['Prism'].highlightElement(this.el.nativeElement);
        }
    };
    CodeHighlighter.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    CodeHighlighter.decorators = [
        { type: Directive, args: [{
                    selector: '[pCode]'
                },] }
    ];
    CodeHighlighter.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    return CodeHighlighter;
}());
export { CodeHighlighter };
var CodeHighlighterModule = /** @class */ (function () {
    function CodeHighlighterModule() {
    }
    CodeHighlighterModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule],
                    exports: [CodeHighlighter],
                    declarations: [CodeHighlighter]
                },] }
    ];
    return CodeHighlighterModule;
}());
export { CodeHighlighterModule };
